cmd/hypershift/info.go

```diff
@@ -0,0 +1,691 @@
+type infoOptions struct {
+	clusterID            string
+	awsProfile           string
+	awsRegion            string
+	privatelinkAccountId string
+	output               string
+	genericclioptions.IOStreams
+}
+
+// InfoClusters aggregates the OCM responses for the customer and management
+// cluster.
+type infoClusters struct {
+	managementCluster *v1.Cluster
+	customerCluster   *v1.Cluster
+}
+
+// Store all clients required to access the 3 different AWS accounts
+type hypershiftAWSClients struct {
+	customerClient    aws.Client
+	managementClient  aws.Client
+	privatelinkClient aws.Client
+}
+
+// Information related to the customer cluster
+type clusterInfo struct {
+	HostedZones       []*route53.HostedZone
+	ResourceRecords   []*route53.ResourceRecordSet
+	Endpoints         []*ec2.VpcEndpoint
+	Subnets           []*ec2.Subnet
+	SubnetRouteTables []*ec2.RouteTable
+}
+
+// Information related to the management cluster
+type managementClusterInfo struct {
+	EndpointConnections []*ec2.VpcEndpointConnection
+	EndpointServices    []*ec2.ServiceDetail
+	LoadBalancers       []*elbv2.LoadBalancer
+}
+
+// Information related to the privatelink account
+type privatelinkInfo struct {
+	HostedZones     []*route53.HostedZone
+	ResourceRecords []*route53.ResourceRecordSet
+	Endpoints       []*ec2.VpcEndpoint
+}
+
+// Aggregates all 3 AWS account informations in one
+type aggregateClusterInfo struct {
+	lock                  sync.Mutex
+	managementClusterInfo *managementClusterInfo
+	clusterInfo           *clusterInfo
+	privatelinkInfo       *privatelinkInfo
+}
+
+func newInfoOptions(streams genericclioptions.IOStreams) *infoOptions {
+	return &infoOptions{
+		IOStreams: streams,
+	}
+}
+
+func (i *infoOptions) complete(cmd *cobra.Command) error {
+	var errMsg string
+	if i.awsProfile == "" {
+		errMsg = "missing argument -p. "
+	}
+	if i.privatelinkAccountId == "" {
+		errMsg += "missing argument -l."
+	}
+	if i.output != "" {
+		if i.output != "graphviz" && i.output != "table" {
+			errMsg += "output must be 'graphviz' or 'table'"
+		}
+	}
+	if errMsg != "" {
+		return fmt.Errorf(errMsg)
+	}
+	return nil
+}
+
+func baseApiUrl(c *v1.Cluster) string {
+	return strings.Replace(strings.Replace(strings.Replace(c.API().URL(), "https://api.", "", 1), c.Name(), "", 1), ":443", "", 1)
+}
+
+func (i *infoOptions) run() error {
+	log.Println("Getting hypershift info for cluster: ", i.clusterID)
+	clusters, err := i.getClusters()
+	if err != nil {
+		return err
+	}
+	if i.awsRegion == "" {
+		i.awsRegion = clusters.customerCluster.Region().ID()
+		log.Println("Set aws region to: ", i.awsRegion)
+	}
+
+	log.Println("Constructing AWS sessions for all accounts")
+	awsSessions, err := i.getAWSSessions(clusters)
+	if err != nil {
+		log.Fatal("Could not construct all AWS sessions: ", err)
+		return err
+	}
+	wg := sync.WaitGroup{}
+	wg.Add(3)
+
+	ai := aggregateClusterInfo{}
+	eC := make(chan error)
```

Comment:
Is there a benefit to using the channels to collect errors?  Not opposed, just curious on the thinking, for my own education.

Reply:
I think it's one of two ways (at least that I know of), to indicate an error during execution of the goroutine.
As the goroutine doesn't return a value at all, there must be another way to check for an error to have occurred.

I think the alternative would be use a single channel to also gather the output and use an aggregated type that wraps the expected return value & an error, so once the value is retrieved the caller can check if an error occurred:

```
type retrievable interface {
	clusterInfo | managementClusterInfo | privatelinkInfo
}

type ChanReturn[T retrievable] struct {
	Value T
	Error error
}
```

Then just use a `chan[ChanReturn]` and pass that to all async functions.

If there is a more synchronous-looking/behaving way I'd be happy to use that - the current mix between output-argument passed in and channel is quite ugly.

Reply:
Thank for the explanation!  This sounds good to me, and I didn't have any issues with it - just trying to learn more myself, thanks!